From c1b7429f23f28d9031cd042740c6e666ffb36aa2 Mon Sep 17 00:00:00 2001
From: Delphine CC Chiu <Delphine_CC_Chiu@wiwynn.com>
Date: Mon, 23 Oct 2023 17:06:54 +0800
Subject: [PATCH 03/20] Support OEM-META write file request for post code
 history

Support write file req when multiple hosts send postcode to BMC
and store its post code history.

Change-Id: I56d303a5bc2f059451ee207a1de591be0d9993ac
Signed-off-by: Delphine CC Chiu <Delphine_CC_Chiu@wiwynn.com>
Upstream-Status: Pending
---
 libpldmresponder/meson.build                  |   9 ++
 meson.build                                   |   4 +
 meson.options                                 |   8 ++
 .../libpldmresponder/oem_meta_file_io.cpp     |  66 +++++++++
 .../libpldmresponder/oem_meta_file_io.hpp     |  49 +++++++
 .../oem_meta_file_io_by_type.hpp              |  32 +++++
 .../oem_meta_file_io_type_post_code.cpp       | 130 ++++++++++++++++++
 .../oem_meta_file_io_type_post_code.hpp       |  53 +++++++
 pldmd/pldmd.cpp                               |   9 ++
 9 files changed, 360 insertions(+)
 create mode 100644 oem/meta/libpldmresponder/oem_meta_file_io.cpp
 create mode 100644 oem/meta/libpldmresponder/oem_meta_file_io.hpp
 create mode 100644 oem/meta/libpldmresponder/oem_meta_file_io_by_type.hpp
 create mode 100644 oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.cpp
 create mode 100644 oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.hpp

diff --git a/libpldmresponder/meson.build b/libpldmresponder/meson.build
index acbb400..c0e3ea7 100644
--- a/libpldmresponder/meson.build
+++ b/libpldmresponder/meson.build
@@ -37,6 +37,7 @@ sources = [
     '../host-bmc/utils.cpp',
     '../host-bmc/dbus/pcie_device.cpp',
     '../host-bmc/dbus/pcie_slot.cpp',
+    '../requester/configuration_discovery_handler.cpp',
     'event_parser.cpp',
 ]
 
@@ -65,6 +66,14 @@ if get_option('oem-ibm').allowed()
     ]
 endif
 
+if get_option('oem-meta').allowed()
+    responder_headers += ['../oem/meta/']
+    sources += [
+        '../oem/meta/libpldmresponder/oem_meta_file_io.cpp',
+        '../oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.cpp',
+    ]
+endif
+
 libpldmresponder = library(
     'pldmresponder',
     sources,
diff --git a/meson.build b/meson.build
index ee237f5..544deec 100644
--- a/meson.build
+++ b/meson.build
@@ -149,6 +149,10 @@ endif
 
 configure_file(output: 'config.h', configuration: conf_data)
 
+if get_option('oem-meta').allowed()
+    add_project_arguments('-DOEM_META', language: ['c', 'cpp'])
+endif
+
 add_project_arguments(
     '-include',
     '@0@'.format(meson.current_build_dir() / 'config.h'),
diff --git a/meson.options b/meson.options
index f49ffc0..aa2f019 100644
--- a/meson.options
+++ b/meson.options
@@ -172,6 +172,14 @@ option(
     description: 'Enable IBM OEM PLDM',
 )
 
+# OEM Meta Options
+option(
+    'oem-meta',
+    type: 'feature',
+    value: 'enabled',
+    description: 'Enable Meta OEM PLDM',
+)
+
 option(
     'oem-ibm-dma-maxsize',
     type: 'integer',
diff --git a/oem/meta/libpldmresponder/oem_meta_file_io.cpp b/oem/meta/libpldmresponder/oem_meta_file_io.cpp
new file mode 100644
index 0000000..763d164
--- /dev/null
+++ b/oem/meta/libpldmresponder/oem_meta_file_io.cpp
@@ -0,0 +1,66 @@
+#include "oem_meta_file_io.hpp"
+
+#include "oem_meta_file_io_type_post_code.hpp"
+#include "xyz/openbmc_project/Common/error.hpp"
+
+#include <libpldm/oem/meta/file_io.h>
+
+#include <filesystem>
+namespace pldm::responder::oem_meta
+{
+
+using namespace sdbusplus::xyz::openbmc_project::Common::Error;
+
+static constexpr auto mctpEndpointInterface =
+    "xyz.openbmc_project.Configuration.MCTPEndpoint";
+
+std::unique_ptr<FileHandler> FileIOHandler::getHandlerByType(uint8_t messageTid,
+                                                             uint8_t fileIOType)
+{
+    switch (fileIOType)
+    {
+        case POST_CODE:
+            return std::make_unique<PostCodeHandler>(
+                messageTid, configurationDescovery->getConfigurations());
+        default:
+            error("Get invalid file io type, FILEIOTYPE={FILEIOTYPE}",
+                  "FILEIOTYPE", fileIOType);
+            break;
+    }
+    return nullptr;
+}
+
+Response FileIOHandler::writeFileIO(pldm_tid_t tid, const pldm_msg* request,
+                                    size_t payloadLength)
+{
+    constexpr uint8_t decodereqlen =
+        PLDM_OEM_META_FILE_IO_WRITE_REQ_MIN_LENGTH + decodeDataMaxLength;
+    alignas(struct pldm_oem_meta_file_io_write_req) unsigned char
+        reqbuf[decodereqlen];
+    auto request_msg = new (reqbuf) pldm_oem_meta_file_io_write_req;
+    auto request_data = static_cast<uint8_t*>(
+        pldm_oem_meta_file_io_write_req_data(request_msg));
+
+    int rc = decode_oem_meta_file_io_write_req(request, payloadLength,
+                                               request_msg, decodereqlen);
+
+    if (rc != 0)
+    {
+        return ccOnlyResponse(request, rc);
+    }
+
+    auto handler = getHandlerByType(tid, request_msg->handle);
+
+    if (!handler)
+    {
+        return ccOnlyResponse(request, PLDM_ERROR_UNSUPPORTED_PLDM_CMD);
+    }
+
+    auto data = std::span(request_data, request_msg->length);
+
+    rc = handler->write(data);
+
+    return ccOnlyResponse(request, rc);
+}
+
+} // namespace pldm::responder::oem_meta
diff --git a/oem/meta/libpldmresponder/oem_meta_file_io.hpp b/oem/meta/libpldmresponder/oem_meta_file_io.hpp
new file mode 100644
index 0000000..d5c173d
--- /dev/null
+++ b/oem/meta/libpldmresponder/oem_meta_file_io.hpp
@@ -0,0 +1,49 @@
+#pragma once
+
+#include "common/utils.hpp"
+#include "oem_meta_file_io_by_type.hpp"
+#include "pldmd/handler.hpp"
+#include "requester/configuration_discovery_handler.hpp"
+
+#include <libpldm/oem/meta/file_io.h>
+
+#include <phosphor-logging/lg2.hpp>
+
+PHOSPHOR_LOG2_USING;
+
+namespace pldm::responder::oem_meta
+{
+
+class FileIOHandler : public CmdHandler
+{
+  public:
+    FileIOHandler(pldm::ConfigurationDiscoveryHandler* configurationDescovery) :
+        configurationDescovery(configurationDescovery)
+    {
+        handlers.emplace(
+            PLDM_OEM_META_FILE_IO_CMD_WRITE_FILE,
+            [this](pldm_tid_t tid, const pldm_msg* request,
+                   size_t payloadLength) {
+                return this->writeFileIO(tid, request, payloadLength);
+            });
+    }
+
+  private:
+    /** @brief Handler for writeFileIO command
+     *
+     *  @param[in] tid - the device tid
+     *  @param[in] request - pointer to PLDM request payload
+     *  @param[in] payloadLength - length of the message
+     *
+     *  @return PLDM response message
+     */
+    Response writeFileIO(pldm_tid_t tid, const pldm_msg* request,
+                         size_t payloadLength);
+
+    std::unique_ptr<FileHandler> getHandlerByType(pldm_tid_t tid,
+                                                  uint8_t fileIOType);
+
+    pldm::ConfigurationDiscoveryHandler* configurationDescovery;
+};
+
+} // namespace pldm::responder::oem_meta
diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_by_type.hpp b/oem/meta/libpldmresponder/oem_meta_file_io_by_type.hpp
new file mode 100644
index 0000000..3f27583
--- /dev/null
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_by_type.hpp
@@ -0,0 +1,32 @@
+#pragma once
+#include <libpldm/oem/meta/file_io.h>
+
+#include <cstdint>
+#include <vector>
+
+namespace pldm::responder::oem_meta
+{
+/**
+ *  @class FileHandler
+ *
+ *  Base class to handle read/write of all OEM file types
+ */
+
+using message = std::span<uint8_t>;
+using configDbusPath = std::string;
+constexpr auto decodeDataMaxLength = 32;
+
+enum pldm_oem_meta_file_io_type : uint8_t
+{
+    POST_CODE = 0x00,
+};
+
+class FileHandler
+{
+  public:
+    virtual int write(const message& data) = 0;
+    virtual int read(struct pldm_oem_meta_file_io_read_resp* data) = 0;
+    virtual ~FileHandler() {}
+};
+
+} // namespace pldm::responder::oem_meta
diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.cpp b/oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.cpp
new file mode 100644
index 0000000..e453af5
--- /dev/null
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.cpp
@@ -0,0 +1,130 @@
+#include "oem_meta_file_io_type_post_code.hpp"
+
+#include <phosphor-logging/lg2.hpp>
+
+#include <utility>
+
+PHOSPHOR_LOG2_USING;
+namespace pldm::responder::oem_meta
+{
+
+using postcode_t = std::tuple<std::vector<uint8_t>, std::vector<uint8_t>>;
+
+void PostCodeHandler::parseObjectPathToGetSlot(uint64_t& slotNum)
+{
+    static constexpr auto slotInterface =
+        "xyz.openbmc_project.Inventory.Decorator.Slot";
+    static constexpr auto slotNumberProperty = "SlotNumber";
+
+    std::vector<std::string> slotInterfaceList = {slotInterface};
+    pldm::utils::GetAncestorsResponse response;
+    bool found = false;
+
+    for (const auto& [configDbusPath, mctpEndpoint] : configurations)
+    {
+        if (mctpEndpoint.EndpointId == tid)
+        {
+            try
+            {
+                response = pldm::utils::DBusHandler().getAncestors(
+                    configDbusPath.c_str(), slotInterfaceList);
+            }
+            catch (const sdbusplus::exception_t& e)
+            {
+                error(
+                    "GetAncestors call failed with, ERROR={ERROR} PATH={PATH} INTERFACE={INTERFACE}",
+                    "ERROR", e, "PATH", configDbusPath.c_str(), "INTERFACE",
+                    slotInterface);
+            }
+
+            // It should be only one of the slot property existed.
+            if (response.size() != 1)
+            {
+                error("Get invalide number of slot property, SIZE={SIZE}",
+                      "SIZE", response.size());
+                return;
+            }
+
+            try
+            {
+                slotNum = pldm::utils::DBusHandler().getDbusProperty<uint64_t>(
+                    std::get<0>(response.front()).c_str(), slotNumberProperty,
+                    slotInterface);
+                found = true;
+            }
+            catch (const sdbusplus::exception_t& e)
+            {
+                error(
+                    "Error getting Names property, ERROR={ERROR} PATH={PATH} INTERFACE={INTERFACE}",
+                    "ERROR", e, "PATH", std::get<0>(response.front()).c_str(),
+                    "INTERFACE", slotInterface);
+            }
+        }
+    }
+
+    if (!found)
+    {
+        throw std::invalid_argument("Configuration of TID not found.");
+    }
+}
+
+int PostCodeHandler::write(const message& postCodeList)
+{
+    static constexpr auto dbusService = "xyz.openbmc_project.State.Boot.Raw";
+    static constexpr auto dbusObj = "/xyz/openbmc_project/state/boot/raw";
+    static constexpr auto AMD_POST_CODE_LENGTH = 4;
+    uint64_t slot;
+
+    try
+    {
+        parseObjectPathToGetSlot(slot);
+    }
+    catch (const std::exception& e)
+    {
+        error("Fail to get the corresponding slot. TID={TID}, ERROR={ERROR}",
+              "TID", tid, "ERROR", e);
+        return PLDM_ERROR;
+    }
+
+    std::string dbusObjStr = dbusObj + std::to_string(slot);
+
+    std::vector<uint8_t> primaryPostCode;
+
+    // Putting list of the bytes together to form a meaningful postcode
+    // AMD platform send four bytes as a post code unit
+    if (std::cmp_greater(
+            std::distance(std::begin(postCodeList), std::end(postCodeList)),
+            sizeof(AMD_POST_CODE_LENGTH)))
+    {
+        error("Get invalid post code length");
+        return PLDM_ERROR;
+    }
+
+    std::for_each(postCodeList.rbegin(), postCodeList.rend(),
+                  [&primaryPostCode](auto postcode) {
+                      primaryPostCode.push_back(postcode);
+                  });
+
+    try
+    {
+        auto& bus = pldm::utils::DBusHandler::getBus();
+        auto method =
+            bus.new_method_call(dbusService, dbusObjStr.c_str(),
+                                "org.freedesktop.DBus.Properties", "Set");
+
+        method.append(
+            dbusService, "Value",
+            std::variant<postcode_t>(postcode_t(primaryPostCode, {})));
+
+        bus.call(method);
+    }
+    catch (const std::exception& e)
+    {
+        error("Set Post code error. ERROR={ERROR}", "ERROR", e);
+        return PLDM_ERROR;
+    }
+
+    return PLDM_SUCCESS;
+}
+
+} // namespace pldm::responder::oem_meta
diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.hpp b/oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.hpp
new file mode 100644
index 0000000..81ff4eb
--- /dev/null
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_type_post_code.hpp
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "common/utils.hpp"
+#include "oem_meta_file_io_by_type.hpp"
+#include "requester/configuration_discovery_handler.hpp"
+
+namespace pldm::responder::oem_meta
+{
+/** @class PostCodeHandler
+ *
+ *  @brief Inherits and implements FileHandler. This class is used
+ *  to store incoming postcode
+ */
+class PostCodeHandler : public FileHandler
+{
+  public:
+    PostCodeHandler(pldm_tid_t tid,
+                    const std::map<std::string, MctpEndpoint>& configurations) :
+        tid(tid), configurations(configurations)
+    {}
+
+    ~PostCodeHandler() = default;
+
+    /** @brief Method to store postcode list
+     *  @param[in] data - post code
+     *  @return  PLDM status code
+     */
+    int write(const message& data);
+
+    /** @brief Method to read postcode list
+     *  @param[in] data - post code
+     *  @return  PLDM status code
+     */
+    int read(struct pldm_oem_meta_file_io_read_resp* data)
+    {
+        (void)data; // Unused
+        return PLDM_ERROR_UNSUPPORTED_PLDM_CMD;
+    }
+
+  private:
+    /** @brief Parse object path to get correspond slot number
+     *  @param[in] slot - slot number
+     */
+    void parseObjectPathToGetSlot(uint64_t& slot);
+
+    /** @brief The terminus ID of the message source*/
+    pldm_tid_t tid = 0;
+
+    /** @brief Get existing configurations with MctpEndpoint*/
+    const std::map<configDbusPath, MctpEndpoint>& configurations;
+};
+
+} // namespace pldm::responder::oem_meta
diff --git a/pldmd/pldmd.cpp b/pldmd/pldmd.cpp
index 3b68896..8f3817c 100644
--- a/pldmd/pldmd.cpp
+++ b/pldmd/pldmd.cpp
@@ -67,6 +67,9 @@ PHOSPHOR_LOG2_USING;
 #ifdef OEM_AMPERE
 #include "oem/ampere/oem_ampere.hpp"
 #endif
+#ifdef OEM_META
+#include "libpldmresponder/oem_meta_file_io.hpp"
+#endif
 
 constexpr const char* PLDMService = "xyz.openbmc_project.PLDM";
 
@@ -275,6 +278,12 @@ int main(int argc, char** argv)
             pldmTransport.getEventSource(), hostEID, instanceIdDb, &reqHandler);
     }
 
+#ifdef OEM_META
+    invoker.registerHandler(
+        PLDM_OEM, std::make_unique<pldm::responder::oem_meta::FileIOHandler>(
+                      configurationDiscovery.get()));
+#endif
+
     auto fruHandler = std::make_unique<fru::Handler>(
         FRU_JSONS_DIR, FRU_MASTER_JSON, pdrRepo.get(), entityTree.get(),
         bmcEntityTree.get());
-- 
2.43.0

